### DISCUSSION TITLE
gem5-SystemC Co-Simulation

### DISCUSSION THREAD

**Discussion by Alibana** [gem5/gem5#1755] [Q&A]
Hello everyone,

I'm new to gem5 but have experience with SystemC. I'm setting up a gem5-SystemC co-simulation and came across this paper, though the examples are limited. I’d appreciate more guidance or a full-system example.
https://grk2767.tu-dresden.de/files/Images/people/chair-cc/publications/1707_Menard_SAMOS.pdf

The Goal:

    Setup: PCIe card with ARM host, 1GiB DRAM, and an accelerator (SystemC).
    Goal: Offload an NN (e.g., LeNet) where matrix multiplications run on the SystemC accelerator, with other operations handled by the ARM host.

I’m trying to use the master/slave_ports examples in gem5/util/tlm, but progress is slow.
Questions:

    Any detailed examples or documentation on the gem5-SystemC co-simulation flow?
    Can gem5’s components (e.g., board, cache) be integrated into the setup to simplify it?
    Is it possible to build the accelerator directly in gem5 instead of SystemC? If so, how?

Thanks in advance for your help!

**Comment by giactra** [comment]
> Any detailed examples or documentation on the gem5-SystemC co-simulation flow?

With respect to the paper/solution you posted, there's a bit more documentation here: https://old.gem5.org/wiki/images/a/a5/Summit2017_systemc.pdf


> Is it possible to build the accelerator directly in gem5 instead of SystemC? If so, how?

Actually some time ago a new gem5+SystemC integration type has been added to gem5: rather than running **gem5 within SystemC** (prior approach you posted), another solution is to run **SystemC within gem5** (basically the SystemC simulation kernel has been implemented within gem5 so you can define SystemC objects natively within the simulator)

You can find examples for both methodologies in the util directory (have a look at the README)

https://github.com/gem5/gem5/tree/stable/util/systemc

**Reply by bsordy** [reply]
If you use the second variant with SystemC within Gem5, is it possible to take a RiscvTimingSimpleCPU() from Gem5 to execute a binary with control information that sends commands to an NPU built in SystemC TLM? If so would the way be via a custom SimObject and memory mapping for instructions or is there a variant via tlm natively from gem5 in the scenario? The Reason for this is that a finished cpu model is to be used to control the systemc simulation. Many thanks for a reply!
